cc/dual_net/tf_dual_net.cc

```diff
@@ -33,80 +37,79 @@ using tensorflow::TensorShape;
 
 namespace minigo {
 
-TfDualNet::TfDualNet(const std::string& graph_path) : graph_path_(graph_path) {
-  GraphDef graph_def;
-
-  // If we can't find the specified graph, try adding a .pb extension.
-  auto* env = Env::Default();
-  if (!env->FileExists(graph_path_).ok()) {
-    auto alt_path = absl::StrCat(graph_path_, ".pb");
-    if (env->FileExists(alt_path).ok()) {
-      std::cerr << graph_path << " doesn't exist, using " << alt_path
-                << std::endl;
-      graph_path_ = alt_path;
+namespace {
+class TfDualNet : public DualNet {
+ public:
+  explicit TfDualNet(std::string model_path) : model_path_(model_path) {
+    // If we can't find the specified graph, try adding a .pb extension.
+    auto* env = tensorflow::Env::Default();
+    if (!env->FileExists(model_path).ok()) {
+      model_path = absl::StrCat(model_path, ".pb");
     }
-  }
-
-  TF_CHECK_OK(ReadBinaryProto(env, graph_path_, &graph_def));
 
-  SessionOptions options;
-  options.config.mutable_gpu_options()->set_allow_growth(true);
-  session_.reset(NewSession(options));
-  TF_CHECK_OK(session_->Create(graph_def));
+    tensorflow::GraphDef graph_def;
+    TF_CHECK_OK(tensorflow::ReadBinaryProto(env, model_path, &graph_def));
 
-  inputs_.clear();
-  inputs_.emplace_back(
-      "pos_tensor",
-      Tensor(DT_FLOAT, TensorShape({1, kN, kN, kNumStoneFeatures})));
+    tensorflow::SessionOptions options;
+    options.config.mutable_gpu_options()->set_allow_growth(true);
+    session_.reset(tensorflow::NewSession(options));
+    TF_CHECK_OK(session_->Create(graph_def));
 
-  output_names_.clear();
-  output_names_.push_back("policy_output");
-  output_names_.push_back("value_output");
+    inputs_.emplace_back(
+        "pos_tensor",
+        tensorflow::Tensor(tensorflow::DT_FLOAT,
+                           tensorflow::TensorShape(
+                               {FLAGS_batch_size, kN, kN, kNumStoneFeatures})));
```

Comment:
[I realized I'm answering the question 'Why are you coupling the DualNetFactory to batching?' here. I hope it will still explain my thinking]

Adding a DualNetFactory implementation which directly creates the backend DualNets seems unnecessary. Batching is pointless for 'eval' and 'gtp' mode because they play single games, but it doesn't hurt to run them through batching either. I would like to make 'eval' play 'parallel_games' in a separate CL. This and MLPerf's puzzle require batching as well, leaving gtp as 1 of 4 paths that technically doesn't benefit from batching.

An alternative would be to expose the NewDualNet() function (currently private to factory.cc). It could take a batch_size parameter, and you would pass 'virtual_losses' for gtp. That would make the batch_size less of an implementation detail than the current approach though in my opinion.

Another alternative would be to have DualNet dynamically resize tensors in RunMany to whatever batch size they are asked to run. This would be more general, but unnecessarily complicate the code.